### PR TITLE
Avoid central calls when resolved from file system

### LINKED ISSUE #32675: Avoid querying the central when exact match is needed and the version is available in filesystem
**Description:**

When the versions are available in Dependencies.toml and a non-expired `target/build` is present,  internally it should be considered as `sticky == true, and locking mode == HARD` which means we should match the exact version. If that version is available in the filesystem repos, we should not query the central.

**Steps to Reproduce:**

1. Create a package with a few dependencies and build with `--dump-build-time` flag and note the package resolution time
      This will create the Dependencies.toml and the `target/build` file
2. Re-run the build command and compare the package resolution times

*_Expected Behavior:_*
The package resolution time for the 2nd attempt should be much lesser than the first time since it is expected to resolve from the filesystem and so no central querying should take place.

*_Actual Behavior:_*
Both times are almost equal

**Affected Versions:**
Beta.3-SNAPSHOT

### PR DESCRIPTION
**Pull request by Dilhasha**
## Purpose
Avoid central calls when resolved from file system.

Fixes #32675

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java [hunk 1]
```diff
@@ -242,27 +243,35 @@ private PackageNameResolutionRequest toPackageNameResolutionRequest(List<ImportM
...
-            throw new ProjectException(e.getMessage());
+        // Remove the already resolved requests when the locking mode is hard
+        for (PackageMetadataResponse response : cachedPackages) {
+            if (response.packageLoadRequest().packageLockingMode().equals(PackageLockingMode.HARD)
+                    && response.resolutionStatus().equals(ResolutionResponse.ResolutionStatus.RESOLVED)) {
+                requests.remove(response.packageLoadRequest());
+            }
+        }
+        // Resolve the requests from remote repository if there are unresolved requests
+        if (requests.size() > 0) {
```
Comment (lnash94) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java:254]:
```suggestion
        if (requests.isEmpty()) {
```

Comment (Dilhasha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java:254]: @lnash94 it should be `if (!requests.isEmpty()) {`

Comment (Dilhasha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java]: Fixed with 9113a96f8a8d84844aae0b5b4cb4e4aa61df0f50

**Review by azinneera — APPROVED** [review]

**Review by sameerajayasoma — COMMENTED** [review]
Is there a way to test this change?

**Comment by azinneera** [comment]
> Is there a way to test this change?

We can add a test case to distribution

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java [hunk 2]
```diff
@@ -242,27 +243,35 @@ private PackageNameResolutionRequest toPackageNameResolutionRequest(List<ImportM
...
-        } catch (ConnectionErrorException e) {
-            // ignore connect to remote repo failure
-            // TODO we need to add diagnostics for resolution errors
-        } catch (CentralClientException e) {
-            throw new ProjectException(e.getMessage());
+        // Remove the already resolved requests when the locking mode is hard
+        for (PackageMetadataResponse response : cachedPackages) {
+            if (response.packageLoadRequest().packageLockingMode().equals(PackageLockingMode.HARD)
+                    && response.resolutionStatus().equals(ResolutionResponse.ResolutionStatus.RESOLVED)) {
+                requests.remove(response.packageLoadRequest());
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java:250]: A better approach would be to create a new list here. Modifying the method parameters should be avoided as much as possible.

**Review by sameerajayasoma — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java (hunk 2 above)
Comment (Dilhasha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java]: fixed with 68e2b272179cba27dae47408c84cb84d3e85adbc

**Review by sameerajayasoma — APPROVED** [review]

**Pull request merged**